PV E2E:  dynamically provisioned volume should not create in unmanaged zone

copejon opened the pull request:
**What this PR does / why we need it**:
Adds e2e test to that attempts to provision a volume in an unmanaged zone and fails on success.  This is to catch regressions of #31948.


cc @jeffvance 
```release-note
NONE
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41049)
<!-- Reviewable:end -->

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -121,6 +123,61 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			Expect(err).NotTo(HaveOccurred())
+
+			// Get a list of all zones in the project
+			compSvc := gceCloud.GetComputeService()
```
Is there a helper func to get all zones and/or get a specific zone?

jeffvance commented on test/e2e/volume_provisioning.go:
```diff
@@ -121,6 +123,61 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+
+			// Get a list of all zones in the project
+			compSvc := gceCloud.GetComputeService()
+			call := compSvc.Zones.List(framework.TestContext.CloudConfig.ProjectID)
```
Is there a helper func to get all zones and/or get a specific zone?

jsafrane commented on test/e2e/volume_provisioning.go:
```diff
@@ -26,9 +26,9 @@ import (
-
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/util/sets"
```
this belongs to the chunk of k8s imports above.

jsafrane commented on test/e2e/volume_provisioning.go:
```diff
@@ -121,6 +121,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			unmanagedZones := allZones.Difference(managedZones)
+			// And select one of them at random.
+			if unmanagedZone, popped = unmanagedZones.PopAny(); ! popped {
+				framework.Failf("No unmanaged zones found.")
```
I'd skip the test in this case - it's not a failure.

jsafrane commented on test/e2e/volume_provisioning.go:
```diff
@@ -121,6 +121,56 @@ var _ = framework.KubeDescribe("Dynamic provisioning", func() {
+			// Set an unmanaged zone.
+			sc.Parameters = map[string]string{"zone": unmanagedZone,}
+			sc, err = c.Storage().StorageClasses().Create(sc)
+			defer Ω(c.Storage().StorageClasses().Delete(sc.Name, nil)).Should(Succeed())
```
Please use `Expect` syntax instead of `Ω`.

jsafrane commented:
@k8s-bot ok to test

jsafrane commented:
```
I0208 04:53:19.259] Verifying hack/make-rules/../../hack/verify-gofmt.sh
I0208 04:53:28.013] !!! 'gofmt -s -w' needs to be run on the following files: 
I0208 04:53:28.013] ./test/e2e/volume_provisioning.go
I0208 04:53:28.022] FAILED   hack/make-rules/../../hack/verify-gofmt.sh	9s
```

jsafrane commented:
now it looks fine, @copejon please squash the commits and I'll lgtm it (is it a verb?)

spxtr commented:
/approve

jsafrane commented:
@copejon, hack/verify-gofmt.sh is still not happy, please reformat test/e2e/volume_provisioning.go

copejon committed:
Add test to check disk cannot be provisioned in unmanaged zones

test dyn/prov'd PVC cannot provision a volume in an unmanaged zone

use k8s gce api wrappers

gofmt'd; do skip if no unmanaged zones

jsafrane commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot kops aws e2e test this

jsafrane commented:
/lgtm
